src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php

```diff
@@ -69,20 +70,24 @@ public function decode(array $encodedEnvelope): Envelope
             throw new MessageDecodingFailedException('Encoded envelope does not have a "type" header.');
         }
 
+        ['body' => $body, 'headers' => ['type' => $type]] = $encodedEnvelope;
         $stamps = $this->decodeStamps($encodedEnvelope);
-        $stamps[] = new SerializedMessageStamp($encodedEnvelope['body']);
+        $stamps[] = new SerializedMessageStamp($body, $type);
 
-        $serializerStamp = $this->findFirstSerializerStamp($stamps);
+        $context = ($this->findFirstSerializerStamp($stamps)?->getContext() ?? [])
+            + $this->context;
 
-        $context = $this->context;
-        if (null !== $serializerStamp) {
-            $context = $serializerStamp->getContext() + $context;
+        try {
+            $body = $this->serializer->decode($body, $this->format, $context);
+        } catch (\Throwable $e) {
+            throw new MessageDecodingFailedException($e->getMessage(), $e->getCode(), $e);
         }
 
         try {
-            $message = $this->serializer->deserialize($encodedEnvelope['body'], $encodedEnvelope['headers']['type'], $this->format, $context);
-        } catch (ExceptionInterface $e) {
-            throw new MessageDecodingFailedException('Could not decode message: '.$e->getMessage(), $e->getCode(), $e);
+            $message = $this->serializer->denormalize($body, $type, $this->format, $context);
+        } catch (\Throwable $e) {
+            $stamps[] = new MessageDecodingFailedStamp($e->getMessage());
+            $message = $this->serializer->denormalize($body, \stdClass::class, $this->format, $context);
```

Comment:
not sure it is very common, but what happens if a handler consume messages of class `\stdClass::class`?

FMI, why do we need to separate the two behaviors, when decoding fails and when denormalization fails?

Reply:
- Anything related to handlers does not matter here. We handle the denormalization failure using stamps long before handlers.
- We split decoding and denormalization to match `PhpSerializer` behavior when it allows incomplete classes: an `Envelope` must be generated, message must be a valid object.